### ISSUE TITLE
Implement Mesh()

### ISSUE THREAD

**Issue by arpastrana** [arpastrana/apc524_directional_clustering#11]
A mesh object that can host different vector fields. For every vector field, only one vector can be assigned per mesh face. Our Mesh() should be a subclass of compas.datastructures.Mesh.

**Comment by arpastrana** [comment]
So I comment on `MeshPlus` here (nice name, btw!).

Concretely, what I'm thinking of as next steps is this:

- To have a method called `MeshPlus().vector_fields()` that returns a list with all the possible vector_field names stored on the Mesh. In the case of script `01.py` there is this list with all the names at the very top of the file (`vectorfield_tags`). Other meshes may or may not have the same number of fields and they may not have the same names. What we can consider as probable vector fields names are _all the unique names of the attributes defined at all the faces of a mesh whose value is a sequence (list, tuple, dict) with three float values (e.g. [0.0, 1.0, 2.0])_. If for example, there is one attribute defined at all the faces of a mesh like `mesh.face_attribute(key=0, name="wrong") = "hello"` or  `mesh.face_attribute(key=0, name="wrong") = ["a", None, 0.0]`, these won't be considered as a vector field candidate. Conversely, if `mesh.face_attribute(key=999, name="ok") = (0.0, 1.0, 2.0)`, then "ok" would be part of the output of calling MeshPlus.vector_fields().

- In `01.py` it would be helpful to not have the vector field name as an initial input parameter, and instead first prompt the user with the list of available vector fields stored in the mesh .json file they are feeding in (`perimeter_supported_slab`, for instance), and then require them to type in the name of the vector field they want to actually run the clustering on (`m_1`, now set as default).

- To replace `VectorField.from_meshes()` with `MeshPlus().vector_field()` in the driver code. This method would double as a getter and a setter. The signature would be `MeshPlus().vector_field(name, vector_field=None)`. If vector_field is None, then the function would try to fetch the face attributes and spit a new Vector Field (replacing what `MeshPlus.get_vectorfield` currently does.) If vector_field is a `VectorField()` object, then this will assign the entries of the vector field as attributes to the faces of a mesh. This is analogous in style and signature to what `mesh.face_attribute` does.

- Afterwards, we'd have to use this functions in `01.py`. First to replace `VectorField.from_mesh_faces()` (this function should be deleted eventually, as `MeshPlus.vector_field()` make semantically more sense, I believe). Then, in the section _"Assign clusters back to COMPAS mesh"_  in`01.py`, we can create a new `MeshPlus`, store the resulting clustered vector field (e.g. `new_mesh = MeshPlus()`, `new_mesh.vector_field("original_field_name_{}_{}".format("k", n_clusters), clustered_field)`, to finally do `new_mesh.to_json()` instead of `mesh.to_json()`.

- We would need to test all (or at least most) of this. To get started, we can simply start cut-pasting the tests related to `VectorField.mesh_from_faces()` in `tests/fields`. We would need to add more as necessary to ensure that the functions above are working okay-ish against future changes. 

In summary: `MeshPlus` with two methods: `vector_field(name, vector_field)`, and `vector_fields(name)`, plus driver code mods, plus testing, and docstringing.

Do you think all of this is a good idea? Would you like to implement it?

**Comment by hhhhuiyuan** [comment]
Thanks for your detailed comment! We should definitely consider which `vectorfield_tags` are supported by a certain mesh, which I wasn't aware of before. So this consideration results in adding an extra method in `MeshPlus` returning a list of supported tags, and modifying the CLI. Another future step is implementing `MeshPlus.vectorfield()`, which serves as a getter(similar to `MeshPlus.get_vectorfield`) as well as a setter. And finally, use these functions in 01.py and do some testing! I'll look at implementation today.

**Comment by arpastrana** [comment]
Awesome — thanks, Hui!

**Comment by hhhhuiyuan** [comment]
Extra methods we discussed about are done! Updates live in branch mesh.

While working on MeshPlus, I found a potential bug in Class Field.
There is a block like:
```
if len(value) != self.dimensionality():
msg = "Length of {} is incompatible with the field's dimensionality"
raise ValueError(msg.format(value))
```
to check the compatibility between the input value and the field.
However, when `value` is an object without length, it cause a TypeError, which is unexpected.

**Comment by arpastrana** [comment]
Great; thanks, @hhhhuiyuan! 
What you point out makes sense. How do you think we could fix thos potential bug in `Field()`?

My take is that this is somewhat expected behavior as the current subclass of `Field`, `VectorField`, expects a sequence of floats as input in `VectorField.add_vector(key, vector)` since `Field.dimensionality() > 1`.

However, if a field was a hypotethical `ScalarField()`, then it should be taking single floats, not a sequence of them. But I believe this could be handled directly by every child class instead(?).

Looking forward to hearing your thoughts on this.

**Comment by hhhhuiyuan** [comment]
Your points make sense, but it will cause trouble in searching supported vector_field attributes based on my current code if we don't fix this bug and just leave it to child classes of `Field`. I made a detailed comment regarding this issue under your suggestions made on my latest commit in mesh branch. Feel free to check it. 
Plus, for the generality of `Field`, as you mentioned that there could be subclass of it taking single float,  I think we should fix it in `Field` directly.

**Comment by arpastrana** [comment]
Thanks, @hhhhuiyuan. My take is that `len(value) != self.dimensionality()` should raise an error at all times (we can live with the hypothetical `ScalarField` edge case for now since this is not part of the package). This is the expected behavior, and I suggest we leave this as it is. As I commented in f9ee20a8516635e779094d9ca4c318d89b9fbedc, the logic of searching for valid vector fields is specific to `MeshPlus.vector_fields()` at this point in time (`try-except` is ok), or even better, to a different method that pre-computes all vector fields at `MeshPlus` instantiation time. 

What do you think?

**Comment by arpastrana** [comment]
Let's just keep it as it is for now. At this point, I'll recommend just adding docstrings and tests, PR-ing, and merging into `develop`. We can refactor later.

**Comment by arpastrana** [comment]
Closed with #37.

**Comment by arpastrana** [comment]
@hhhhuiyuan, I closed and merged the PR :) Thanks a lot for this. The test file for meshplus was empty...will you still add the tests?

**Comment by hhhhuiyuan** [comment]
yes, I'll add some basic tests, to make the test folder look completed.

**Comment by arpastrana** [comment]
cool!
